tower/src/util/mod.rs

```diff
@@ -82,6 +85,240 @@ pub trait ServiceExt<Request>: tower_service::Service<Request> {
+    /// use tower_util::ServiceExt;
+    ///
+    /// # struct DatabaseService;
+    /// # impl DatabaseService {
+    /// #   fn new(address: &str) -> Self {
+    /// #       DatabaseService  
+    /// #   }
+    /// # }
+    /// #
+    /// struct Error {
+    ///    pub code: u32,
+    ///    pub message: String
+    /// }
+    ///
+    /// # impl Service<u32> for DatabaseService {
+    /// #   type Response = String;
+    /// #   type Error = Error;
+    /// #   type Future = futures_util::future::Ready<Result<String, Error>>;
+    /// #
+    /// #   fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+    /// #       Poll::Ready(Ok(()))
+    /// #   }
+    /// #
+    /// #   fn call(&mut self, request: u32) -> Self::Future {
+    /// #       futures_util::future::ready(Ok(String::new()))
+    /// #   }
+    /// # }
+    /// #
+    /// fn main() {
+    ///     // A service returning Result<_, Error>
+    ///     let service = DatabaseService::new("127.0.0.1:8080");
+    ///
+    ///     // Map the error into a new error
+    ///     let mut new_service = service.map_err(|err| err.code);
+    ///
+    ///     async {
+    ///         let id = 13;
+    ///         let code = new_service.call(id).await.unwrap_err();
+    ///     };
+    /// }
+    /// ```
+    fn map_err<F, Error>(self, f: F) -> MapErr<Self, F>
+    where
+        Self: Sized,
+        F: FnOnce(Self::Error) -> Error + Clone,
+    {
+        MapErr::new(self, f)
+    }
+
+    /// Composes a function *in front of* the service.
+    ///
+    /// This adapter produces a new service that passes each value through the
+    /// given function `f` before sending it to `self`.
+    ///
+    /// # Example
+    /// ```
+    /// # use std::convert::TryFrom;
+    /// # use std::task::{Poll, Context};
+    /// # use tower_service::Service;
+    /// use tower_util::ServiceExt;
+    ///
+    /// # struct DatabaseService;
+    /// # impl DatabaseService {
+    /// #   fn new(address: &str) -> Self {
+    /// #       DatabaseService  
+    /// #   }
+    /// # }
+    /// #
+    /// # impl Service<u32> for DatabaseService {
+    /// #   type Response = String;
+    /// #   type Error = u8;
+    /// #   type Future = futures_util::future::Ready<Result<String, u8>>;
+    /// #
+    /// #   fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+    /// #       Poll::Ready(Ok(()))
+    /// #   }
+    /// #
+    /// #   fn call(&mut self, request: u32) -> Self::Future {
+    /// #       futures_util::future::ready(Ok(String::new()))
+    /// #   }
+    /// # }
+    /// #
+    /// fn main() {
+    ///     // A service taking an u32 as a request
+    ///     let service = DatabaseService::new("127.0.0.1:8080");
+    ///
+    ///     // Map the request into a new request
+    ///     let mut new_service = service.with(|id_str: &str| id_str.parse().unwrap());
+    ///
+    ///     async {
+    ///         let id = "13";
```

Comment:
similarly, i think we need a `u32` type annotation. also, it looks like this is a string literal?
```suggestion
    ///         let id: u32 = 13;
```